apps/hal9/utils.py

```diff
@@ -0,0 +1,329 @@
+    tool_calls = getattr(response_message, 'tool_calls', None)
+
+    if not tool_calls:
+        print("No tool calls found.")
+        return
+
+    # Iterate over the tool calls and extract relevant information.
+    for tool_call in tool_calls:
+        function_name = tool_call.function.name
+        try:
+            arguments = ast.literal_eval(tool_call.function.arguments)
+        except AttributeError as e:
+            print(f"Error accessing arguments: {e}")
+            continue
+        # Convert arguments into a string format for logging or execution.
+        args_str = ', '.join(f"{k}={repr(v)}" for k, v in arguments.items())
+
+        # Add all the functions into the exec context
+        context = {}
+        for func in functions:
+            context[func.__name__] = func
+        
+        # Prepare the code string to execute
+        code_to_exec = f"result = {function_name}({args_str})"
+
+        # Execute the code with exec(), but ensure proper error handling.
+        try:
+            exec(code_to_exec, context)
+            return context['result']
+        except Exception as e:
+            print(f"Error executing function '{function_name}': {e}")
+            raise
+
+def stream_print(text: str):
+       for char in text:
+        print(char, end="", flush=True)
+        time.sleep(0.02)
+
+def insert_tool_message(messages, model_response, tool_result):
+    tool_calls = model_response.choices[0].message.tool_calls
+
+    if tool_calls:
+      for tool_call in tool_calls:
+        messages.append({
+          "role": "assistant",
+          "tool_calls": [{
+            "id": tool_call.id,
+            "type": "function",
+            "function": {
+              "arguments": tool_call.function.arguments,
+              "name": tool_call.function.name,
+            },
+          }]
+        })
+        function_args = json.loads(tool_call.function.arguments, strict=False)
+
+        tool_content = json.dumps({**function_args, "response": str(tool_result)})
+
+        messages.append({
+            "role": "tool",
+            "content": tool_content,
+            "tool_call_id": tool_call.id
+        })
+
+def is_url(prompt):
+  result = urllib.parse.urlparse(prompt)
+  return all([result.scheme, result.netloc])
+
+def download_file(url):
+    # Create the .storage directory if it doesn't exist
+    if not os.path.exists("./.storage"):
+        os.makedirs("./.storage")
+    filename = url.split("/")[-1]
+    modified_filename = f"./.storage/.{filename}"
+    
+    response = requests.get(url)
+    
+    if response.status_code == 200:
+        with open(modified_filename, 'wb') as file:
+            file.write(response.content)
+    else:
+        print(f"Failed to download the file. Status code: {response.status_code}")
+
+def generate_embeddings(text, model, client_type):
+    client = get_client(client_type)
+    response = client.embeddings.create(
+    input=text,
+    model=model)
```

Comment:
Did we consider generating the embeddings in parallel? Or is this something the OpenAI client already does?

Reply:
Gotcha, thanks!